### ISSUE TITLE
Incorrectly creates firewall rules for local delivery

### ISSUE THREAD

**Issue by mhuxtable** [debops/ansible-postfix#92]
Firewall rules permitting external access to the mail server are created when the `local` capability is the only capability enabled. This is an unexpected and unnecessary relaxation of host security.

# Situation

I am configuring Postfix on a host for which output from local daemons/processes must be sent to an external destination. This mail will be delivered directly by Postfix using MX lookup for the destination MTA. The default `nullmailer` daemon is inadequate because there is no local relay available to relay the mail, nor is it necessary in this circumstance as the host can deliver the mail directly.

The most common example will be local daemons and processes (e.g. `cron`) sending mail to `root`, which will be redirected to an external destination via an alias in `/etc/aliases`.

# Configuration

The host is present in the `[debops_service_postfix]` inventory group. Postfix is configured as follows:

```
postfix: [ 'local' ]
postfix_local_aliases:
  root: [ 'me@some-external-place.com' ]
postfix_mydomain: '{{ ansible_fqdn }}'
```

# Expected behaviour

Mail delivered to root on the local host is routed to the external address defined in aliases. No access is permitted to the Postfix instance running on the host from any origin outside the host.

# Actual behaviour

The example command `mail -s "Test Email" root` on the host correctly delivers email to the external destination defined in aliases.

However, the `iptables` output includes the following rules in the `INPUT` chain:

```
ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:25
ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:465
ACCEPT     tcp  --  0.0.0.0/0            0.0.0.0/0            tcp dpt:587
```

Postfix is configured with `inet_interfaces = loopback-only`, so external access is still rejected as public interfaces are not bound by the process. However, this behaviour depends on process-specific configuration to avoid binding the public interfaces on the host. I would prefer the firewall rules were not created, so as to provide defence-in-depth and enable a minimal set of firewall rules on the host to ease auditing.

# Analysis

According to the [capabilities documentation](https://docs.debops.org/en/latest/ansible/roles/ansible-postfix/docs/defaults-configuration.html#postfix-capabilities) and an analysis of the relevant code (see below), I do not expect the firewall rules permitting public access to be created.

The [ferm configuration](https://github.com/debops/ansible-postfix/blob/d80dd4113189a5dcb4d46f6c3d8388677d180c7e/defaults/main.yml#L561-L566) in the Postfix role, extract copied below, indicates firewall rules allowing public access should only be present when the configured capabilities match `network AND (mx OR submission)`.

```
    when: '{{ ("network" in postfix_capabilities and
               ("mx" in postfix_capabilities or
                "submission" in postfix_capabilities)) | bool }}'
    delete: '{{ ("network" not in postfix_capabilities or
                 ("mx" not in postfix_capabilities and
                  "submission" not in postfix_capabilities)) | bool }}'
```

When running the playbook, the conditionals are evaluated correctly - observe below Ansible `debug` output:

```
    "postfix_ferm_dependent_rules": [
        {
            "accept_any": true, 
            "delete": true, 
            "dport": [
                "smtp"
            ], 
            "name": "smtp", 
            "role": "postfix", 
            "saddr": [], 
            "type": "accept", 
            "weight": "40", 
            "when": false
        }, 
        (smtps and submission rules elided - they are identical to above save for the dport field)
    ]
```

Note `delete: true` is present for each rule.

Although it appears this should be actioned correctly for rules created by `debops.ferm` as `filter-input.d` rules ([see code](https://github.com/debops/ansible-ferm/blob/c41465cd0eeae57630b331d90a5bc46f88047150/tasks/main.yml#L107-L128)), the same `when` and `delete` parameters are not respected by the ferm role when creating `rules.d` rules for services running on the local host ([see code](https://github.com/debops/ansible-ferm/blob/c41465cd0eeae57630b331d90a5bc46f88047150/tasks/main.yml#L86-L115)).

This may be a bug in the `debops.ferm` role instead. I am reporting it here as I have no evidence this is a more general issue beyond this Postfix role.

**Comment by drybjed** [comment]
This is an excellent analysis of the issue!

This is not a bug in the `debops.ferm`, but rather an outdated code in the `debops.postfix` role. For some time now I've been updating the roles to not use the `when` and `delete` parameters in tasks, but rather use the `state` parameter, similar to how Ansible `state:` directive works. Therefore the reason the rules are created, is that the `debops.ferm` role no longer recognizes the `delete` parameter in the configuration. `debops.postfix` configuration should be updated to reflect that fact.

The other reason for the open firewall is the use of `accept_any: True` parameter (which is the default in `debops.ferm` now anyway). If it would have been `False`, the rules would still be created, but firewall would open access to everyone by default.

Right now I'm in the process of redesigning the `debops.postfix` role from scratch. It will be much smaller in scope, but much more "composable" with other roles - you will be able to configure Postfix from other roles as you see fit. I plan to drop the Postfix capabilities from this role and move the respective functionality (like `auth`, `client`, etc.) to separate Ansible roles which will be activated as needed. The PR for the new role should be up in a couple of days for testing.

Keep in mind that the `debops.postfix` role in its current form was designed when Postfix was included by default in the DebOps `common.yml` role. At the time, blocking access to SMTP service by default was a good idea. Since then, `nullmailer` became the default and Postfix can be now enabled on demand by Ansible groups. Because of that I plan to allow firewall access to everyone by default, with an option to limit the subnets that can access the SMTP servers, or reverse the setting and block all by default and only allow specific subnets.

Right now what you can do, is to change the `ferm` Postfix configuration through Ansible inventory. Since it's passed through to the `debops.ferm` using default Ansible variables, overriding it in Ansible inventory should pass your own configuration instead of the default one. Can you try that and let me know if that works?

**Comment by mhuxtable** [comment]
> This is not a bug in the `debops.ferm`, but rather an outdated code in the `debops.postfix` role.

That's good to know! I suspected that was the case as I saw similar comments in other `debops.postfix` issues relating to this role being implemented before the current API across the Debops project was adopted.

> Right now what you can do, is to change the `ferm` Postfix configuration through Ansible inventory. [...] Can you try that and let me know if that works?

That's what I ended up doing. I am pleased to say it works correctly and the iptables rules do not get created in that case.

For the benefit of anyone coming here from Google, I added the following line to `ansible/inventory/host_vars/<hostname>/postfix.yml`:

```
postfix_ferm_dependent_rules: []
```

although it could equally go into `group_vars/<some_inventory_groupname>/postfix.yml` or `group_vars/all/postfix.yml` or some other inventory variable location according your requirements.

Note: I don't recall explicitly testing the rules get deleted after setting the above and re-running it on a host with the erroneous rules present from a previous application of this Postfix role. My understanding of the `ferm` role suggests it should tidy up rules which are no longer specified in the yml spec. However, I may have manually deleted the rule specs in `/etc/ferm/rules.d/140_*`, so anyone applying this fix is advised to check the rules do indeed disappear after a re-run of the relevant playbook.

> I plan to allow firewall access to everyone by default, with an option to limit the subnets that can access the SMTP servers, or reverse the setting and block all by default and only allow specific subnets.

I can think of scenarios in which both configuration options would be useful, including the current one in which `nullmailer` is not quite enough and I don't need Postfix exposed to the world. So being explicit about this in the documentation and configuration parameters would be a useful addition.

> Right now I'm in the process of redesigning the `debops.postfix` role from scratch

I will refrain from digging any deeper on this issue and use the inventory workaround for now. I would be happy to test your new role as and when it becomes available, and if I can help further with any documentation around the behaviour of these parameters, do let me know!
